diff --git a/assets/css/wcmp-admin-styles.css b/assets/css/wcmp-admin-styles.css
index f7607f5..fa0cbe0 100755
--- a/assets/css/wcmp-admin-styles.css
+++ b/assets/css/wcmp-admin-styles.css
@@ -18,7 +18,53 @@
 }
 
 /*
- * Shipment summary
+ * Shipment overview
+ */
+.wcmp__order__track-trace {
+  word-break: break-all;
+}
+
+/*
+ * Shipment overview: Track & Trace column
+ */
+.wcmp__barcodes {
+  overflow: hidden;
+  text-overflow: ellipsis;
+  white-space: nowrap;
+}
+
+.wcmp__box {
+  background-color: white;
+  border: 1px solid #ccc;
+  box-shadow: 0 0 10px rgba(0, 0, 0, 0.15);
+  padding: 10px;
+  position: absolute;
+  z-index: 999;
+  border-radius: 3px;
+}
+
+.wcmp__box .form-row {
+  margin: 0.3rem 0 !important;
+}
+
+/*
+ * Shipment overview: Print offset dialog
+ */
+.wcmp__offset-dialog {
+  position: absolute;
+}
+
+.wcmp__offset-dialog .button {
+  display: none;
+  margin-top: 4px;
+}
+
+.wcmp__offset-dialog__icon {
+  vertical-align: middle;
+}
+
+/*
+ * Shipment summary (overview and single order)
  */
 .wcmp__shipment-summary, .wcmp__shipment-options {
   display: inline;
@@ -38,27 +84,6 @@
   margin-right: 4px;
 }
 
-/*
- * Track & Trace in shipment overview
- */
-.wcmp__barcodes {
-  overflow: hidden;
-  text-overflow: ellipsis;
-  white-space: nowrap;
-}
-
-/*
- * The print offset dialog
- */
-.wcmp__offset-dialog {
-  position: absolute;
-  background-color: white;
-  padding: 6px;
-  width: 100px;
-  border: 1px solid #ccc;
-  margin-left: -100px;
-}
-
 /*
  * Shipment options form styling
  */
@@ -76,15 +101,6 @@
   vertical-align: middle;
 }
 
-.wcmp__shipment-options__form,
-.wcmp__shipment-summary__list {
-  border: 1px solid #ccc;
-  background: white;
-  position: absolute;
-  z-index: 999;
-  padding: 10px;
-}
-
 .wcmp__shipment-summary__list {
   width: auto;
 }
@@ -120,11 +136,13 @@
 .wcmp__spinner--loading .wcmp__spinner__loading,
 .wcmp__spinner--success .wcmp__spinner__success,
 .wcmp__spinner--failed .wcmp__spinner__failed {
+  margin-left: 4px;
   display: block !important;
 }
 
 .button > .wcmp__spinner {
-  margin-left: 4px;
+  margin-left: 0;
+  transition: all 0.3s ease;
 }
 
 .button + .wcmp__spinner, input + .wcmp__spinner {
@@ -255,6 +273,10 @@
   display: block;
 }
 
+.wcmp__d--inline-block {
+  display: inline-block;
+}
+
 .wcmp__d--flex {
   display: flex;
 }
@@ -303,6 +325,6 @@
   width: 280px;
 }
 
-.wcmp__order__track-trace {
-  word-break: break-all;
+.wcmp__close-button {
+  cursor: pointer;
 }
diff --git a/assets/js/wcmp-admin.js b/assets/js/wcmp-admin.js
index 6b52093..67dbc37 100755
--- a/assets/js/wcmp-admin.js
+++ b/assets/js/wcmp-admin.js
@@ -10,11 +10,11 @@
  * @property {{export: String, print: String, export_print: String}} wcmp.bulk_actions
  * @property {String} wcmp.download_display
  * @property {String} wcmp.nonce
- * @property {String} wcmp.offset_icon
+ * @property {Object.<String, String>} wcmp.strings
  */
 
 /* eslint-disable-next-line max-lines-per-function */
-jQuery(function ($) {
+jQuery(function($) {
   /**
    * @type {Boolean}
    */
@@ -22,7 +22,9 @@ jQuery(function ($) {
 
   var selectors = {
     offsetDialog: '.wcmp__offset-dialog',
-    offsetDialogInput: '.wcmp__offset-dialog__offset',
+    offsetDialogButton: '.wcmp__offset-dialog__button',
+    offsetDialogClose: '.wcmp__offset-dialog__close',
+    offsetDialogInputOffset: '.wcmp__offset-dialog__offset',
     printQueue: '.wcmp__print-queue',
     printQueueOffset: '.wcmp__print-queue__offset',
     saveShipmentSettings: '.wcmp__shipment-settings__save',
@@ -62,6 +64,8 @@ jQuery(function ($) {
      */
     $(selectors.offsetDialog + ' button').click(printOrder);
 
+    $(selectors.offsetDialogClose).click(hideOffsetDialog);
+
     /**
      * Show and enable options when clicked.
      */
@@ -83,7 +87,7 @@ jQuery(function ($) {
     /**
      * Add offset dialog when address labels option is selected.
      */
-    $('select[name=\'action\'], select[name=\'action2\']').change(showOffsetDialog);
+    $('select[name=\'action\'], select[name=\'action2\']').change(showBulkOffsetDialog);
 
     /**
      * Single actions click. The .wc_actions .single_wc_actions for support wc > 3.3.0.
@@ -363,6 +367,11 @@ jQuery(function ($) {
 
     $(rows.join(', ')).addClass('wcmp__loading');
 
+    if (!order_ids.length) {
+      alert(wcmp.strings.no_orders_selected);
+      return;
+    }
+
     switch (action) {
       /**
        * Export orders.
@@ -377,7 +386,6 @@ jQuery(function ($) {
       case wcmp.bulk_actions.print:
         printLabel({
           order_ids: order_ids,
-          offset: askForPrintPosition ? $(selectors.offsetDialogInput).val() : 0,
         });
         break;
 
@@ -477,7 +485,11 @@ jQuery(function ($) {
         exportToMyParcel.bind(button)();
         break;
       case wcmp.actions.get_labels:
-        printLabel.bind(button)();
+        if (askForPrintPosition && !$(button).hasClass('wcmp__offset-dialog__button')) {
+          showOffsetDialog.bind(button)();
+        } else {
+          printLabel.bind(button)();
+        }
         break;
       case wcmp.actions.add_return:
         myparcelbe_modal_dialog(order_ids, 'return');
@@ -485,113 +497,116 @@ jQuery(function ($) {
     }
   }
 
-  function showOffsetDialog() {
-    var selectedAction = $(this).val();
-    var offsetDialog = $(selectors.offsetDialog);
-
-    if ((selectedAction === 'wcmp_print' || selectedAction === 'wcmp_export_print') && askForPrintPosition) {
-      var insert_position = $(this).attr('name') === 'action' ? 'top' : 'bottom';
+  /**
+   * Show the offset dialog before printing.
+   *
+   * @param {String?} position - To position the dialog `left` or `right` relative to the bound element.
+   * @param {String?} context - Context in which the dialog was created. Ex. 'bulk'.
+   */
+  function showOffsetDialog(position, context) {
+    position = position || 'left';
 
-      offsetDialog
-        .attr('style', 'clear:both') /* reset styles */
-        .insertAfter('div.tablenav.' + insert_position)
-        .show();
+    var parent = this;
+    var offsetDialog = $(selectors.offsetDialog);
+    var dialogButton = $(selectors.offsetDialogButton);
+    var parentOffset = $(parent).offset();
 
-      /* make sure button is not shown */
-      offsetDialog.find('button').hide();
-      /* clear input */
-      offsetDialog.find('input').val('');
+    /**
+     * Position it to the bottom left or right of the clicked button.
+     */
+    if (position === 'left') {
+      offsetDialog.css({
+        left: parentOffset.left - offsetDialog.width(),
+        top: parentOffset.top,
+      });
     } else {
-      offsetDialog
-        .appendTo('body')
-        .hide();
+      offsetDialog.css(parentOffset);
     }
-  }
 
-  function printOrder() {
-    var dialog = $(this).parent();
+    dialogButton.attr('href', parent.href);
 
-    /* set print variables */
-    var order_ids = [dialog.find('input.order_id').val()];
-    var offset = dialog.find(selectors.offsetDialogInput).val();
+    /**
+     * Reset input(s).
+     */
+    offsetDialog.find('input').val(0);
 
-    /* hide dialog */
-    dialog.hide();
+    /**
+     * Make sure button is not shown and there is no input listener to update it if context is bulk.
+     */
+    if (context === 'bulk') {
+      dialogButton.hide();
+      $(selectors.offsetDialogInputOffset).off('blur update change', onUpdateOffset);
+    } else {
+      dialogButton.show();
+      $(selectors.offsetDialogInputOffset).on('blur update change', onUpdateOffset);
+    }
 
-    /* print labels */
-    printLabel({
-      order_ids: order_ids,
-      offset: offset,
-    });
+    /**
+     * Finally, show the dialog.
+     */
+    offsetDialog.slideDown();
   }
 
   /**
-   * Place offset dialog at mouse tip.
+   * Hide the offset dialog and remove the input listener.
+   *
+   * @param {Event?} event - Click event if called from a button.
    */
-  function showContextualOffsetDialog(order_ids, event) {
-    var offsetDialog = $(selectors.offsetDialog);
-
-    offsetDialog
-      .show()
-      .appendTo('body')
-      .css({
-        top: event.pageY,
-        left: event.pageX,
-      });
-
-    offsetDialog.find('button')
-      .show()
-      .data('order_id', order_ids);
-
-    /* clear input */
-    offsetDialog.find('input').val('');
+  function hideOffsetDialog(event) {
+    if (event) {
+      event.preventDefault();
+    }
 
-    offsetDialog.append('<input type="hidden" class="order_id"/>');
-    $(selectors.offsetDialog + ' input.order_id').val(order_ids);
+    $(selectors.offsetDialogInputOffset).off('blur update change', onUpdateOffset);
+    $(selectors.offsetDialog).slideUp();
   }
 
   /**
-   * @param {Element} button - The button that was clicked.
-   * @param {Boolean} display - To display or not to display.
+   * On changing the offset value in the dialog, update the offset parameter in the dialog button's href attribute.
    */
-  function showButtonSpinner(button, display) {
-    if (display) {
-      var buttonImage = $(button).find(selectors.orderActionImage);
-      buttonImage.hide();
-      $(button).parent()
-        .find(selectors.spinner)
-        .insertAfter(buttonImage)
-        .show();
+  function onUpdateOffset() {
+    var dialogButton = $(selectors.offsetDialogButton);
+    var hasOffset = dialogButton.attr('href').indexOf('offset=') > -1;
+    var newOffset = this.value;
+
+    if (hasOffset) {
+      dialogButton.attr('href', dialogButton.attr('href').replace(/([?&]offset=)\d*/, '$1' + newOffset));
     } else {
-      $(button).parent()
-        .find(selectors.spinner)
-        .hide();
-      $(button).find(selectors.orderActionImage)
-        .show();
+      dialogButton.attr('href', dialogButton.attr('href') + '&offset=' + newOffset);
     }
-  }
+  };
 
   /**
-   * @param {Element} action - The action that was clicked.
-   * @param {Boolean} display - To display or not to display.
+   * Show the offset dialog for bulk options that allow it.
    */
-  function showBulkSpinner(action, display) {
-    var submit_button = $(action)
-      .parent()
-      .find('.button.action');
-
-    if (display) {
-      $(selectors.bulkSpinner)
-        .insertAfter(submit_button)
-        .show();
-    } else {
-      $(selectors.bulkSpinner).hide();
+  function showBulkOffsetDialog() {
+    if ([wcmp.bulk_actions.print, wcmp.bulk_actions.export_print].indexOf(this.value) === -1) {
+      hideOffsetDialog();
+      return;
     }
+
+    showOffsetDialog.bind(this)('right', 'bulk');
+  }
+
+  function printOrder() {
+    var dialog = $(this).parent();
+
+    /* set print variables */
+    var order_ids = [dialog.find('input.order_id').val()];
+    var offset = dialog.find(selectors.offsetDialogInputOffset).val();
+
+    /* hide dialog */
+    dialog.hide();
+
+    /* print labels */
+    printLabel({
+      order_ids: order_ids,
+      offset: offset,
+    });
   }
 
   /* export orders to MyParcel via AJAX */
   function exportToMyParcel(order_ids, print) {
-    var offset = askForPrintPosition ? $(selectors.offsetDialogInput).val() : 0;
     var url;
     var data;
 
@@ -605,7 +620,7 @@ jQuery(function ($) {
       data = {
         action: wcmp.actions.export,
         request: wcmp.actions.add_shipments,
-        offset: offset,
+        offset: getPrintOffset(),
         order_ids: order_ids,
         print: print,
         _wpnonce: wcmp.nonce,
@@ -686,10 +701,36 @@ jQuery(function ($) {
   }
 
   /**
+   * Open given pdf link. Depending on the link it will be either downloaded or viewed. Refreshes the original window.
+   *
    * @param {String} pdfUrl - The url of the created pdf.
+   * @param {Boolean?} waitForOnload - Wait for onload to refresh the original window. Refreshes immediately if false.
+   *
    */
-  function openPdf(pdfUrl) {
-    window.open(pdfUrl);
+  function openPdf(pdfUrl, waitForOnload) {
+    var pdfWindow = window.open(pdfUrl, '_blank');
+
+    if (waitForOnload) {
+      /*
+       * When the pdf window is loaded reload the main window. If we reload earlier the track & trace code won't be
+       * ready yet and can't be shown.
+       */
+      pdfWindow.onload = function() {
+        window.location.reload();
+      };
+    } else {
+      /* For when there is no onload event or there is no need to wait. */
+      window.location.reload();
+    }
+  }
+
+  /**
+   * Get the offset from the offset dialog if it's present. Otherwise return 0.
+   *
+   * @returns {Number}
+   */
+  function getPrintOffset() {
+    return parseInt(askForPrintPosition ? $(selectors.offsetDialogInputOffset).val() : 0);
   }
 
   /* Request MyParcel BE labels */
@@ -706,6 +747,7 @@ jQuery(function ($) {
         data: Object.assign({
           action: wcmp.actions.export,
           request: wcmp.actions.get_labels,
+          offset: getPrintOffset(),
           _wpnonce: wcmp.nonce,
         }, data),
       };
@@ -713,7 +755,6 @@ jQuery(function ($) {
 
     request.afterDone = function(response) {
       openPdf(response);
-      window.location.reload();
     };
 
     if (wcmp.download_display === 'download') {
@@ -727,15 +768,7 @@ jQuery(function ($) {
         url = wcmp.ajax_url + '?' + $.param(request.data);
       }
 
-      var pdfWindow = window.open(url, '_blank');
-
-      /**
-       * When the pdf window is loaded reload the main window. If we reload earlier the track & trace code won't be
-       * ready yet and can't be shown.
-       */
-      pdfWindow.onload = function() {
-        window.location.reload();
-      };
+      openPdf(url, true);
     }
   }
 
diff --git a/includes/admin/class-wcmp-admin.php b/includes/admin/class-wcmp-admin.php
index 7b295df..831c528 100644
--- a/includes/admin/class-wcmp-admin.php
+++ b/includes/admin/class-wcmp-admin.php
@@ -2,9 +2,9 @@
 
 use MyParcelNL\Sdk\src\Adapter\DeliveryOptions\AbstractDeliveryOptionsAdapter as DeliveryOptions;
 use MyParcelNL\Sdk\src\Factory\DeliveryOptionsAdapterFactory;
-use MyParcelNL\Sdk\src\Factory\ShipmentOptionsAdapterFactory;
 use WPO\WC\MyParcelBE\Compatibility\WC_Core as WCX;
 use WPO\WC\MyParcelBE\Compatibility\Order as WCX_Order;
+use WPO\WC\MyParcelBE\Entity\SettingsFieldArguments;
 
 if (! defined('ABSPATH')) {
     exit;
@@ -45,7 +45,7 @@ class WCMP_Admin
             add_action("admin_footer", [$this, "bulk_actions"]);
         }
 
-        add_action("woocommerce_after_account_orders", [$this, "renderOffsetDialog"]);
+        add_action("admin_footer", [$this, "renderOffsetDialog"]);
 
         /**
          * Orders page
@@ -82,8 +82,7 @@ class WCMP_Admin
     }
 
     /**
-     * @param      $order
-     * @param bool $hide
+     * @param WC_Order $order
      *
      * @throws Exception
      */
@@ -108,7 +107,7 @@ class WCMP_Admin
             <div class="wcmp__shipment-summary">
                 <?php $this->showDeliveryOptionsForOrder($order); ?>
                 <a class="wcmp__shipment-summary__show"><span class="wcmp__encircle wcmp__shipment-summary__show">i</span></a>
-                <div class="wcmp__shipment-summary__list"
+                <div class="wcmp__box wcmp__shipment-summary__list"
                      data-loaded=""
                      data-shipment_id="<?php echo $last_shipment_id; ?>"
                      data-order_id="<?php echo $order_id; ?>"
@@ -126,7 +125,7 @@ class WCMP_Admin
                 '<a href="#" class="wcmp__shipment-options__show">%s &#x25BE;</a>',
                 __("Details", "woocommerce-myparcelbe")
             ); ?>
-            <div class="wcmp__shipment-options__form" style="display: none;">
+            <div class="wcmp__box wcmp__shipment-options__form" style="display: none;">
                 <a class="wcmp__d--flex">
                     <?php include('views/html-order-shipment-options.php'); ?>
                 </a>
@@ -215,22 +214,51 @@ class WCMP_Admin
      * @access public
      * @return void
      */
-    public function renderOffsetDialog()
+    public function renderOffsetDialog(): void
     {
+        if (! WCMP()->setting_collection->isEnabled(WCMP_Settings::SETTING_ASK_FOR_PRINT_POSITION)) {
+            return;
+        }
+
+        $field = [
+            "name"              => "offset",
+            "class"             => ["wcmp__d--inline-block"],
+            "input_class"        => ["wcmp__offset-dialog__offset"],
+            "type"              => "number",
+            "label"             => __("Labels to skip", "woocommerce-myparcelbe"),
+            "custom_attributes" => [
+                "step" => "1",
+                "min"  => "0",
+                "max"  => "4",
+                "size" => "2",
+            ],
+        ];
+
+        $class = new SettingsFieldArguments($field);
         ?>
+
         <div
-            class="wcmp__offset-dialog"
-            style="display:none;">
-            <?php _e("Labels to skip", "woocommerce-myparcelbe"); ?>: <input
-                type="text"
-                size="2"
-                class="wcmp__offset-dialog__offset"> <img
-                src="<?php echo WCMP()->plugin_url() . '/assets/img/print-offset-icon.png'; ?>"
-                class="wcmp__offset-dialog__icon"
-                style="vertical-align: middle;">
-            <button
-                class="button"
-                style="display:none; margin-top: 4px"><?php _e("Print", "woocommerce-myparcelbe"); ?></button>
+            class="wcmp wcmp__box wcmp__offset-dialog"
+            style="display: none;">
+            <div class="wcmp__offset-dialog__inner wcmp__d--flex">
+                <div>
+                    <?php woocommerce_form_field($field["name"], $class->getArguments(false), ""); ?>
+
+                    <img
+                        src="<?php echo WCMP()->plugin_url() . "/assets/img/print-offset-icon.png"; ?>"
+                        alt="<?php implode(", ", WCMP_Export::DEFAULT_POSITIONS) ?>"
+                        class="wcmp__offset-dialog__icon"/>
+                    <div>
+                        <a
+                            href="#"
+                            class="wcmp__action wcmp__offset-dialog__button button">
+                            <?php _e("Print", "woocommerce-myparcelbe"); ?>
+                            <?php WCMP_Admin::renderSpinner(); ?>
+                        </a>
+                    </div>
+                </div>
+                <div class="wcmp__close-button dashicons dashicons-no-alt wcmp__offset-dialog__close"></div>
+            </div>
         </div>
         <?php
     }
diff --git a/includes/admin/class-wcmp-export.php b/includes/admin/class-wcmp-export.php
index 22fc2d7..a26705d 100644
--- a/includes/admin/class-wcmp-export.php
+++ b/includes/admin/class-wcmp-export.php
@@ -166,7 +166,7 @@ class WCMP_Export
 
         $dialog  = $_REQUEST["dialog"] ?? null;
         $print   = $_REQUEST["print"] ?? null;
-        $offset  = $_REQUEST["offset"] ?? null;
+        $offset  = (bool) $_REQUEST["offset"] ?? 0;
         $request = $_REQUEST["request"];
 
         $order_ids    = $this->sanitize_posted_array($_REQUEST["order_ids"] ?? []);
@@ -413,13 +413,10 @@ class WCMP_Export
         WCMP_Log::add("Shipment IDs: " . implode(", ", $shipment_ids));
 
         try {
-            $api       = $this->init_api();
-            $positions = self::DEFAULT_POSITIONS;
+            $api = $this->init_api();
 
-            if (! empty($offset) && is_numeric($offset)) {
-                // positions are defined on landscape, but paper is filled portrait-wise
-                $positions = array_slice(self::DEFAULT_POSITIONS, $offset);
-            }
+            // positions are defined on landscape, but paper is filled portrait-wise
+            $positions = array_slice(self::DEFAULT_POSITIONS, $offset % 4);
 
             $display = WCMP()->setting_collection->getByName(WCMP_Settings::SETTING_DOWNLOAD_DISPLAY) === "display";
             $api->getShipmentLabels($shipment_ids, $positions, $display);
@@ -1318,17 +1315,15 @@ class WCMP_Export
     }
 
     /**
-     * @param $order_ids
-     * @param $shipment_ids
-     * @param $offset
+     * @param array $order_ids
+     * @param array $shipment_ids
+     * @param int   $offset
      *
      * @return array
      * @throws Exception
      */
-    private function printLabels($order_ids, $shipment_ids, $offset)
+    private function printLabels(array $order_ids, array $shipment_ids, int $offset)
     {
-        $offset = ! empty($offset) && is_numeric($offset) ? $offset % 4 : 0;
-
         if (! empty($shipment_ids)) {
             $return = $this->downloadOrGetUrlOfLabels(
                 $shipment_ids,
diff --git a/includes/class-wcmp-assets.php b/includes/class-wcmp-assets.php
index 3de0a0f..1601bab 100644
--- a/includes/class-wcmp-assets.php
+++ b/includes/class-wcmp-assets.php
@@ -63,7 +63,7 @@ class WCMP_Assets
                 "wcmp-admin",
                 "wcmp",
                 [
-                    "api_url" => WCMP_Data::API_URL,
+                    "api_url"                => WCMP_Data::API_URL,
                     "actions"                => [
                         "export"        => WCMP_Export::EXPORT,
                         "add_return"    => WCMP_Export::ADD_RETURN,
@@ -83,8 +83,9 @@ class WCMP_Assets
                     "ask_for_print_position" => WCMP()->setting_collection->isEnabled(
                         WCMP_Settings::SETTING_ASK_FOR_PRINT_POSITION
                     ),
-                    "offset_icon"            => WCMP()->plugin_url() . "/assets/img/print-offset-icon.png",
-                    "offset_label"           => __("Labels to skip", "woocommerce-myparcelbe"),
+                    "strings"                => [
+                        "no_orders_selected" => __("You have not selected any orders!", "woocommerce-myparcelbe"),
+                    ],
                 ]
             );
 
diff --git a/includes/vendor/myparcelnl/sdk/src/Helper/MyParcelCollection.php b/includes/vendor/myparcelnl/sdk/src/Helper/MyParcelCollection.php
index 16de105..449c5c6 100644
--- a/includes/vendor/myparcelnl/sdk/src/Helper/MyParcelCollection.php
+++ b/includes/vendor/myparcelnl/sdk/src/Helper/MyParcelCollection.php
@@ -287,8 +287,9 @@ class MyParcelCollection extends Collection
              * Loop through the returned ids and add each consignment id to a consignment.
              */
             foreach ($request->getResult('data.ids') as $responseShipment) {
-                $consignments      = clone $this->getConsignmentsByReferenceId($responseShipment['reference_identifier']);
-                $newConsignments[] = $consignments->pop()->setConsignmentId($responseShipment['id']);
+                $consignments      = $this->getConsignmentsByReferenceId($responseShipment['reference_identifier']);
+                $consignment       = clone $consignments->pop();
+                $newConsignments[] = $consignment->setConsignmentId($responseShipment['id']);
             }
         }
 
